cpp/src/arrow/array/array_nested.cc

```diff
@@ -611,9 +611,10 @@ const std::shared_ptr<Array>& StructArray::field(int i) const {
   return boxed_fields_[i];
 }
 
-std::shared_ptr<Array> StructArray::GetFieldByName(const std::string& name) const {
+const std::shared_ptr<Array>& StructArray::GetFieldByName(const std::string& name) const {
+  static std::shared_ptr<Array> null;
```

Comment:
Having to add a static like this might be reason enough to continue returning a value instead of a `&`.

Reply:
Agreed. The refcount cost is also not terribly annoying since we're also doing a hash lookup, presumably.

Reply:
A static like this is also an implicit mutex so it has the same synchronization cost as a refcount

Reply:
reverted